// nl-marshal/lib/optional.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bool_1 = __importDefault(require("./bool"));
const ISerializer_1 = __importDefault(require("./ISerializer"));
class OptionalSerializer extends ISerializer_1.default {
    constructor(serializer) {
        super();
        this.serializer = serializer;
    }
    toJSON(value) {
        if (value === null || value === undefined)
            return null;
        return this.serializer.toJSON(value);
    }
    fromJSON(value) {
        if (value === null)
            return null;
        return this.serializer.toJSON(value);
    }
    toBuffer(value) {
        const missing = value === null || value === undefined;
        const preres = bool_1.default.toBuffer(!missing);
        if (missing)
            return preres;
        return Buffer.concat([preres, this.serializer.toBuffer(value)]);
    }
    readFromBuffer(buffer, offset = 0) {
        const { res: provided, newOffset: from } = bool_1.default.readFromBuffer(buffer, offset);
        if (!provided)
            return { res: null, newOffset: from };
        return this.serializer.readFromBuffer(buffer, from);
    }
}
exports.OptionalSerializer = OptionalSerializer;
function optional(serializer) {
    return new OptionalSerializer(serializer);
}
exports.default = optional;
